odl/space/fspace.py

```diff
@@ -359,192 +481,416 @@ def element(self, fcall=None, fapply=None):
+            if n == 1:
+                return x
+            elif n % 2 == 0:
+                x *= x
+                return ipow_posint(x, n // 2)
+            else:
+                tmp = x.copy()
+                x *= x
+                ipow_posint(x, n // 2)
+                x *= tmp
+                return x
+
+        def power_call_out_of_place(x):
+            """The power out-of-place evaluation function."""
+            if p == int(p) and p >= 1:
+                return pow_posint(x_call_oop(x), int(p))
+            else:
+                return x_call_oop(x) ** p
 
-    def __str__(self):
-        """Return ``str(self)``."""
-        inner_repr = '{}'.format(self.domain)
-        if not isinstance(self.range, RealNumbers):
-            inner_repr += ', {}'.format(self.range)
+        def power_call_in_place(x, out):
+            """The power in-place evaluation function."""
+            x_call_ip(x, out)
+            if p == int(p) and p >= 1:
+                return ipow_posint(out, int(p))
+            else:
+                out **= p
+                return out
 
-        return 'FunctionSpace({})'.format(inner_repr)
+        out._call_out_of_place = power_call_out_of_place
+        out._call_in_place = power_call_in_place
+        out._call_has_out = out._call_out_optional = True
+        return out
 
     @property
     def element_type(self):
-        """ `FunctionSpaceVector` """
+        """`FunctionSpaceVector`"""
         return FunctionSpaceVector
 
 
 class FunctionSpaceVector(FunctionSetVector, LinearSpaceVector):
+
     """Representation of a `FunctionSpace` element."""
 
-    def __init__(self, fspace, fcall=None, fapply=None):
+    def __init__(self, fspace, fcall):
         """Initialize a new instance.
 
         Parameters
         ----------
         fspace : `FunctionSpace`
             The set of functions this element lives in
-        fcall : `callable`, optional
+        fcall : `callable`
             The actual instruction for out-of-place evaluation.
             It must return an `FunctionSet.range` element or a
             `numpy.ndarray` of such (vectorized call).
-        fapply : `callable`, optional
-            The actual instruction for in-place evaluation.
-            Its first argument must be the `FunctionSet.range`
-            element or the array of such (vectorization) to which the
-            result is written.
-
-        *At least one of the arguments ``fcall`` and ``fapply`` must
-        be provided.*
         """
         if not isinstance(fspace, FunctionSpace):
             raise TypeError('function space {!r} not a `FunctionSpace` '
                             'instance.'.format(fspace))
 
-        super().__init__(fspace, fcall, fapply)
+        FunctionSetVector.__init__(self, fspace, fcall)
+        LinearSpaceVector.__init__(self, fspace)
+
+    # Some additional magic methods not defined for arbitrary linear spaces
```

Comment:
It's the operations with scalars which fail, right. The in-place ones work since they use `lincomb` (therefore they're not overridden), out-of-place and `r` versions fail since the default upstream implementation bails out `if other not in space`.

We could move this code to `LinearSpaceVector` like this:
```python
def __add__(self, other):
        """Return ``self + other``."""
        # Instead of using __iadd__ we duplicate code here for performance
        if other in self.space:
            tmp = self.space.element()
            return self.space.lincomb(1, self, 1, other, out=tmp)
        elif other in self.space.field:
            one = getattr(self.space, 'one', None)
            if one is None:
                return NotImplemented
            else:
                # other --> other * space.one()
                tmp = self.space.one()
                self.space.lincomb(other, tmp, out=tmp)
                return self.space.lincomb(1, self, 1, tmp, out=tmp)                
        else:
            return NotImplemented
```
Upside of this is that any new space relying on a `one` element to implement this functionality will work out of the box.